plugins/thunderbolt/fu-thunderbolt-device.c

```diff
@@ -0,0 +1,645 @@
+						 NULL,
+						 error);
+
+	if (os == NULL)
+		return FALSE;
+
+	nwritten = 0;
+	fw_size = g_bytes_get_size (blob_fw);
+	fu_device_set_progress_full (FU_DEVICE (self), nwritten, fw_size);
+
+	do {
+		g_autoptr(GBytes) fw_data = NULL;
+
+		fw_data = g_bytes_new_from_bytes (blob_fw,
+						  nwritten,
+						  fw_size - nwritten);
+
+		n = g_output_stream_write_bytes (os,
+						 fw_data,
+						 NULL,
+						 error);
+		if (n < 0)
+			return FALSE;
+
+		nwritten += n;
+		fu_device_set_progress_full (FU_DEVICE (self), nwritten, fw_size);
+
+	} while (nwritten < fw_size);
+
+	if (nwritten != fw_size) {
+		g_set_error_literal (error,
+				     FWUPD_ERROR,
+				     FWUPD_ERROR_WRITE,
+				     "Could not write all data to nvmem");
+		return FALSE;
+	}
+
+	return g_output_stream_close (os, NULL, error);
+}
+
+static FuFirmware *
+fu_thunderbolt_device_prepare_firmware (FuDevice *device,
+					GBytes *fw,
+					FwupdInstallFlags flags,
+					GError **error)
+{
+	FuThunderboltDevice *self = FU_THUNDERBOLT_DEVICE (device);
+	g_autoptr(FuThunderboltFirmware) firmware = fu_thunderbolt_firmware_new ();
+	g_autoptr(FuThunderboltFirmware) firmware_old = fu_thunderbolt_firmware_new ();
+	g_autoptr(GBytes) controller_fw = NULL;
+	g_autoptr(GError) error_local = NULL;
+	g_autoptr(GFile) nvmem = NULL;
+
+	/* parse */
+	if (!fu_firmware_parse (FU_FIRMWARE (firmware), fw, flags, error))
+		return NULL;
+
+	/* get current NVMEM */
+	nvmem = fu_thunderbolt_device_find_nvmem (self, TRUE, error);
+	if (nvmem == NULL)
+		return NULL;
+	controller_fw = g_file_load_bytes (nvmem, NULL, NULL, error);
+	if (!fu_firmware_parse (FU_FIRMWARE (firmware_old), controller_fw, flags, error))
+		return NULL;
+	if (fu_thunderbolt_firmware_is_host (firmware) !=
+	    fu_thunderbolt_firmware_is_host (firmware_old)) {
+		g_set_error (error,
+			     FWUPD_ERROR,
+			     FWUPD_ERROR_INVALID_FILE,
+			     "incorrect firmware mode, got %s, expected %s",
+			     fu_thunderbolt_firmware_is_host (firmware) ? "host" : "device",
+			     fu_thunderbolt_firmware_is_host (firmware_old) ? "host" : "device");
+		return NULL;
+	}
+	if (fu_thunderbolt_firmware_get_vendor_id (firmware) !=
+	    fu_thunderbolt_firmware_get_vendor_id (firmware_old)) {
+		g_set_error (error,
+			     FWUPD_ERROR,
+			     FWUPD_ERROR_INVALID_FILE,
+			     "incorrect device vendor, got 0x%04x, expected 0x%04x",
+			     fu_thunderbolt_firmware_get_vendor_id (firmware),
+			     fu_thunderbolt_firmware_get_vendor_id (firmware_old));
+		return NULL;
+	}
+	if (fu_thunderbolt_firmware_get_device_id (firmware) !=
+	    fu_thunderbolt_firmware_get_device_id (firmware_old)) {
+		g_set_error (error,
+			     FWUPD_ERROR,
+			     FWUPD_ERROR_INVALID_FILE,
+			     "incorrect device type, got 0x%04x, expected 0x%04x",
+			     fu_thunderbolt_firmware_get_device_id (firmware),
+			     fu_thunderbolt_firmware_get_device_id (firmware_old));
+		return NULL;
+	}
+	if ((flags & FWUPD_INSTALL_FLAG_FORCE) == 0) {
```

Comment:
I would say we may as well just move this to the top of the function and let force override anything, not just the model ID check?

Reply:
do we ever want to allow installing a host firmware on a device? I'd argue we just let the user blow their feet off... Maybe VID and PID makes sense to ignore tho?

Reply:
The only circumstance that this was needed I know of was a mid-development model ID string change.  Maybe you're right - we shouldn't let people shoot themselves in the foot until we know another real need for it.